File: bigseqkit/concat.go
package bigseqkit

import (
	"ignis/driver/api"
	"ignis/executor/api/ipair"
)

type SeqKitConcatOptions struct {
	inner ConcatOptions
}

type ConcatOptions struct {
	Config    KitConfig
	Full      *bool
	Separator *string
}

func (this *ConcatOptions) setDefaults() *ConcatOptions {
	this.Config.setDefaults()
	setDefault(&this.Full, false)
	setDefault(&this.Separator, "|")

	return this
}

func (this *SeqKitConcatOptions) Config(v *SeqKitConfig) *SeqKitConcatOptions {
	this.inner.Config = v.inner
	return this
}

func (this *SeqKitConcatOptions) Seed(v bool) *SeqKitConcatOptions {
	this.inner.Full = &v
	return this
}

func (this *SeqKitConcatOptions) Separator(v string) *SeqKitConcatOptions {
	this.inner.Separator = &v
	return this
}

func prepareConcat(input *api.IDataFrame[string], opts *ConcatOptions, id string) (*api.IDataFrame[ipair.IPair[string, string]], error) {
	libprepare, err := api.AddParam(libSource("ConcatPrepare"), "opts", OptionsToString(*opts))
	if err != nil {
		return nil, err
	}
	libprepare, err = api.AddParam(libprepare, "id", id)
	if err != nil {
		return nil, err
	}
	return api.MapPartitions[string, ipair.IPair[string, string]](input, libprepare)
}

func Concat(inputA *api.IDataFrame[string], inputB *api.IDataFrame[string], o *SeqKitConcatOptions) (*api.IDataFrame[string], error) {
	if o == nil {
		o = &SeqKitConcatOptions{}
	}
	opts := o.inner
	opts.setDefaults()

	p1, err := prepareConcat(inputA, &opts, "1")
	if err != nil {
		return nil, err
	}
	p2, err := prepareConcat(inputB, &opts, "2")
	if err != nil {
		return nil, err
	}

	u, err := p1.Union(p2, false, nil)
	if err != nil {
		return nil, err
	}

	grouped, err := api.GroupByKey[string, string](api.ToPair[string, string](u), nil)
	if err != nil {
		return nil, err
	}

	join, err := api.AddParam(libSource("ConcatJoin"), "opts", OptionsToString(opts))
	if err != nil {
		return nil, err
	}

	return api.Flatmap[ipair.IPair[string, []string], string](grouped.FromPair(), join)
}

File: bigseqkit-py/bigseqkit/concat.py
from bigseqkit.helper import _setDefault, _libSource, _config, _optionsToString, _parseKargs, SeqKitConfig, IDataFrame


class SeqKitConcatOptions:

    def __init__(self):
        self.__inner = ConcatOptions()

    def config(self, v: SeqKitConfig):
        self.__inner.Config = _config(v)

    def full(self, v: bool):
        self.__inner.Full = v

    def separator(self, v: str):
        self.__inner.Separator = v

    def _prepareConcat(self, inputA: IDataFrame, id: str) -> IDataFrame:
        libprepare = _libSource("ConcatPrepare")\
            .addParam("opts", _optionsToString(self.__inner))\
            .addParam("id", id)
        return inputA.mapPartitions(libprepare)
    def _run(self, inputA: IDataFrame, inputB: IDataFrame, **kwargs):
        opts = self.__inner
        _parseKargs(opts, kwargs)
        opts.setDefaults()

        p1 = self._prepareConcat(inputA, "1")
        p2 = self._prepareConcat(inputB, "2")

        u = p1.union(p2, preserveOrder=False)
        grouped = u.toPair().groupByKey()

        join = _libSource("SeqTransform").addParam("opts", _optionsToString(opts))
        return grouped.flatmap(join)

class ConcatOptions:

    def __init__(self):
        self.Config = None  # KitConfig
        self.Full = None  # bool
        self.Separator = None # str

    def setDefaults(self):
        _setDefault(self, "Config", _config(SeqKitConfig())).setDefaults()
        _setDefault(self, "Full", False)
        _setDefault(self, "Separator", "|")


def concat(inputA: IDataFrame, inputB: IDataFrame, o: SeqKitConcatOptions = None, **kwargs):
    if o is None:
        o = SeqKitConcatOptions()
    return o._run(inputA, inputB, **kwargs)
